asv_bench/benchmarks/omnisci/benchmarks.py

```diff
@@ -28,21 +28,62 @@
     trigger_import,
     get_benchmark_shapes,
 )
+from ..utils.common import random_state
+import numpy as np
 
 
 class TimeJoin:
-    param_names = ["shape", "how"]
+    param_names = ["shape", "how", "is_equal_keys"]
     params = [
         get_benchmark_shapes("omnisci.TimeJoin"),
         ["left", "inner"],
+        [True, False],
     ]
 
-    def setup(self, shape, how):
-        self.df1 = generate_dataframe(ASV_USE_IMPL, "int", *shape, RAND_LOW, RAND_HIGH)
-        self.df2 = generate_dataframe(ASV_USE_IMPL, "int", *shape, RAND_LOW, RAND_HIGH)
+    def setup(self, shape, how, is_equal_keys):
+        self.df1, self.df2 = (
+            generate_dataframe(
+                ASV_USE_IMPL,
+                "int",
+                *frame_shape,
+                RAND_LOW,
+                RAND_HIGH,
+                cache_prefix=f"{i}-th_frame_to_join",
+            )
+            for i, frame_shape in enumerate((shape, shape))
+        )
+
+        if is_equal_keys:
+            # When the frames have default indices to join on: RangeIndex(frame_length),
+            # OmniSci backend performs join on the internal meta-column called 'rowid'.
+            # There is a bug in the engine that makes such joins fail. To avoid joining
+            # on the meta-column we explicitly specify a non-default index to join on.
+            # https://github.com/modin-project/modin/issues/3442
+            # Generating a new object for every index to avoid shared index objects:
+            self.df1.index = np.arange(1, len(self.df1) + 1)
+            self.df2.index = np.arange(1, len(self.df2) + 1)
```

Comment:
this feels like a bug which should be handled in the omnisci wrapper, not in the benchmark

Reply:
To what wrapper do you refer? This is a bug in the <i>OmniSci</i> backend, for which we're doing this workaround on the <i>OmniSci</i> benchmark.

For now, there is no way to perform and so to time the join of two frames on their default indices.

Reply:
I refer to Modin code that utilizes OmniSci engine.
It should have the workaround to the issue, maybe by creating a new object if a default index is used.

Any user could be facing this obscure error, and users should not be expected to know how to deal with this type of issues.

Reply:
Agree. But I believe that this should be done as a part of the fix for the actual bug: a temporary patch that sets non-default indices until the problem is fixed at the OmniSci part. Do you want me to do this before this PR? 

Anyway, even if this workaround will be moved to the backend part, this will still make sense to use it in the bench: since the workaround is quite expensive (it uses 'set_index' which is defaulting to pandas), we would like to measure two scenarios: actual 'join' time (with workaround) and potential time (do workaround at the setup and do not measure its time)

Reply:
I mean, this is not even a fix or workaround that we might put into the OmniSci backend in Modin, this is purely a hack to make 'join' work and be able to bench it.

The only way I see of removing these lines from the bench is to make a <i>proper</i> fix for this case in OmniSci.

Reply:
Makes sense. Please open a proper issue, though - you're currently referring to a one that does not describe the reasons of why this fails, but rather points to some by-product of the real bug, after that I would object no more to this hack.

Reply:
Created two issues: in OmniSci repo (intel-ai/omniscidb#214) and in Modin (#3740), changed the linked issue in the in-code comment.